File: packages/python/src/Seguranca/Ataques/AtaqueCiphertextEstatistico.py
"""Estatística só do CIPHERTEXT (não do keystream): distribuição de bytes
(qui-quadrado), teste de runs nos bits e autocorrelação lag-1. Um ciphertext
de cifra sólida deve parecer ruído mesmo com plaintext variado.
"""

import math

from ..ResultadoAtaque import ResultadoAtaque
from ..Util import random_bytes


class AtaqueCiphertextEstatistico:
    def __init__(self, amostras: int = 200, tamanho_texto: int = 64, z_limite: float = 4):
        self.amostras = amostras
        self.tamanho_texto = tamanho_texto
        self.z_limite = z_limite

    def nome(self) -> str:
        return "Estatística do ciphertext (chi²/runs/autocorrelação)"

    def executar(self, alvo) -> ResultadoAtaque:
        contagem = [0] * 256
        bytes_ = []

        for _ in range(self.amostras):
            texto = random_bytes(self.tamanho_texto).decode("latin-1")
            token = alvo.encrypt(texto)
            campos = alvo.decompor(alvo.base64url_decode(token[len(alvo.prefixo()) :]))
            for b in campos["ciphertext"]:
                contagem[b] += 1
                bytes_.append(b)

        total = len(bytes_)
        esperado = total / 256
        chi2 = 0.0
        for b in range(256):
            d = contagem[b] - esperado
            chi2 += (d * d) / esperado

        # runs test nos bits do ciphertext
        uns = 0
        bits = []
        for b in bytes_:
            for k in range(7, -1, -1):
                bit = (b >> k) & 1
                bits.append(bit)
                uns += bit
        n = len(bits)
        pi = uns / n
        runs = 1
        for i in range(1, n):
            if bits[i] != bits[i - 1]:
                runs += 1
        esperado_runs = 2 * n * pi * (1 - pi)
        desvio_runs = 2 * math.sqrt(2 * n) * pi * (1 - pi)
        z_runs = abs(runs - esperado_runs) / desvio_runs if desvio_runs > 0 else 0.0

        # autocorrelação lag-1
        media = sum(bytes_) / total
        num = 0.0
        den = 0.0
        for i in range(total):
            den += (bytes_[i] - media) ** 2
        for i in range(total - 1):
            num += (bytes_[i] - media) * (bytes_[i + 1] - media)
        autocorr = num / den if den > 0 else 0.0

        problemas = []
        if chi2 > 330:
            problemas.append(f"qui-quadrado={chi2:.1f} (>330 suspeito)")
        if z_runs > self.z_limite:
            problemas.append(f"runs z={z_runs:.2f}")
        if abs(autocorr) > 0.1:
            problemas.append(f"autocorrelação={autocorr:.3f}")

        vulneravel = len(problemas) > 0

        return ResultadoAtaque(
            self.nome(),
            vulneravel,
            "media" if vulneravel else "info",
            (
                "; ".join(problemas)
                if vulneravel
                else f"qui-quadrado={chi2:.1f} sobre {total} bytes, runs z={z_runs:.2f}, "
                f"autocorrelação={autocorr:.3f} - todos dentro do esperado"
            ),
            {"chi2": chi2, "runs_z": z_runs, "autocorrelacao": autocorr},
        )

File: packages/dart/lib/Seguranca/Ataques/AtaqueCiphertextEstatistico.dart
import 'dart:math';

import 'package:criptografia/Seguranca/AlvoCriptografico.dart';
import 'package:criptografia/Seguranca/Ataque.dart';
import 'package:criptografia/Seguranca/ResultadoAtaque.dart';
import 'package:criptografia/Seguranca/Util.dart';

/// Estatística só do CIPHERTEXT (não do keystream): distribuição de bytes
/// (qui-quadrado), teste de runs nos bits e autocorrelação lag-1. Um ciphertext
/// de cifra sólida deve parecer ruído mesmo com plaintext variado.
class AtaqueCiphertextEstatistico extends Ataque {
  final int _amostras;
  final int _tamanhoTexto;
  final double _zLimite;

  AtaqueCiphertextEstatistico({
    this._amostras = 200,
    this._tamanhoTexto = 64,
    this._zLimite = 4,
  });

  @override
  String nome() => 'Estatística do ciphertext (chi²/runs/autocorrelação)';

  @override
  ResultadoAtaque executar(AlvoCriptografico alvo) {
    final contagem = List<int>.filled(256, 0);
    final bytes = <int>[];

    for (var s = 0; s < _amostras; s++) {
      final texto = randomBytes(_tamanhoTexto);
      final token = alvo.encrypt(texto);
      final campos = alvo.decompor(
        alvo.base64urlDecode(token.substring(alvo.prefixo().length)),
      );
      for (final b in campos['ciphertext']!) {
        contagem[b]++;
        bytes.add(b);
      }
    }

    final total = bytes.length;
    final esperado = total / 256;
    var chi2 = 0.0;
    for (var b = 0; b < 256; b++) {
      final d = contagem[b] - esperado;
      chi2 += (d * d) / esperado;
    }

    // runs test nos bits do ciphertext
    var uns = 0;
    final bits = <int>[];
    for (final b in bytes) {
      for (var k = 7; k >= 0; k--) {
        final bit = (b >> k) & 1;
        bits.add(bit);
        uns += bit;
      }
    }
    final n = bits.length;
    final pi = uns / n;
    var runs = 1;
    for (var i = 1; i < n; i++) {
      if (bits[i] != bits[i - 1]) runs++;
    }
    final esperadoRuns = 2 * n * pi * (1 - pi);
    final desvioRuns = 2 * sqrt(2 * n) * pi * (1 - pi);
    final zRuns = desvioRuns > 0
        ? (runs - esperadoRuns).abs() / desvioRuns
        : 0.0;

    // autocorrelação lag-1
    final media = bytes.reduce((a, b) => a + b) / total;
    var num = 0.0;
    var den = 0.0;
    for (var i = 0; i < total; i++) {
      den += pow(bytes[i] - media, 2).toDouble();
    }
    for (var i = 0; i < total - 1; i++) {
      num += (bytes[i] - media) * (bytes[i + 1] - media);
    }
    final autocorr = den > 0 ? num / den : 0.0;

    final problemas = <String>[];
    if (chi2 > 330) {
      problemas.add('qui-quadrado=${chi2.toStringAsFixed(1)} (>330 suspeito)');
    }
    if (zRuns > _zLimite) {
      problemas.add('runs z=${zRuns.toStringAsFixed(2)}');
    }
    if (autocorr.abs() > 0.1) {
      problemas.add('autocorrelação=${autocorr.toStringAsFixed(3)}');
    }

    final vulneravel = problemas.isNotEmpty;

    return ResultadoAtaque(
      nome(),
      vulneravel,
      vulneravel ? Severidade.media : Severidade.info,
      vulneravel
          ? problemas.join('; ')
          : 'qui-quadrado=${chi2.toStringAsFixed(1)} sobre $total bytes, '
                'runs z=${zRuns.toStringAsFixed(2)}, '
                'autocorrelação=${autocorr.toStringAsFixed(3)} - todos '
                'dentro do esperado',
      {'chi2': chi2, 'runs_z': zRuns, 'autocorrelacao': autocorr},
    );
  }
}
